packages/ics721/src/ibc_packet_receive.rs

```diff
@@ -154,190 +241,49 @@ pub(crate) fn receive_ibc_packet(
             // - using class id as salt for instantiating nft contract guarantees a) predictable address and b) uniqueness
             // for this salt must be of length 32 bytes, so we use sha256 to hash class id
             let mut hasher = Sha256::new();
-            hasher.update(voucher.class.id.as_bytes());
+            hasher.update(local_class_id.as_bytes());
             let salt = hasher.finalize().to_vec();
 
-            get_instantiate2_address(
-                deps.as_ref(),
-                env.contract.address.as_str(),
-                &salt,
-                cw721_code_id,
-            )
-        } else {
-            // This should never happen, as we must have at least 1 of the above actions
-            Err(ContractError::InvalidTransferNoAction)
+            get_instantiate2_address(deps, env.contract.address.as_str(), &salt, cw721_code_id)
         }?;
 
-        generate_receive_callback_msg(
-            deps.as_ref(),
-            &cloned_data,
+        Ok(generate_receive_callback_msg(
+            deps,
+            data,
             receive_callback_data,
             receive_callback_addr,
             nft_contract.to_string(),
-        )
-    } else {
-        None
-    };
-
-    let incoming_proxy_msg =
-        get_incoming_proxy_msg(deps.storage, packet.clone(), cloned_data.clone())?;
-    let submessage = action_aggregator.into_submessage(
-        env.contract.address,
-        receiver,
-        callback_msg,
-        incoming_proxy_msg,
-    )?;
```

Comment:
Here is all the main logic: submessage itself contains a list of messags for:
- action aggregrator: creates a callback msg with a list of more messages for create or redeemining NFTs
- optional incoming proxy handling
- optional callback handling